markdown/util.py

```diff
@@ -45,60 +44,269 @@
 """
 
 
-class _BlockLevelElements(set):
-    # ----------------------------------
-    # Methods common to `list` and `set`
-    # ----------------------------------
-    def copy(self) -> set[str]:
-        return _BlockLevelElements(super().copy())
+# TODO: Raise errors from list methods in the future.
+# Later, remove this class entirely and use a regular set.
+class _BlockLevelElements:
+    def __init__(self, elements: list[str]) -> None:
+        self._list = elements.copy()
+        self._set = set(self._list)
+
+    def __add__(self, other: list[str]) -> list[str]:
+        warnings.warn(
+            "Using block level elements as a list is deprecated, use it as a set instead.",
+            DeprecationWarning,
+        )
+        # Using `+` means user expects a list back.
+        return self._list + other
```

Comment:
Methods with no side-effects return a `list` or `set` directly, not a copy of our hybrid instance. The reason is we assume that as soon as a user explicitly uses it as a list (or set), they will only use it as a list (or set), so no need to continue providing both interfaces.

Reply:
Not necessarily. The same object is likely to be  accessed by multiple extensions. It is reasonable to assume that some newly updated extensions will assume a set and some older (not yet updated) extensions will assume a list. Whatever solution we chose needs to continue to work with both throughout the life of the `Markdown` class instance.

Reply:
Yeah I thought about that, though since these methods return a new object (that's what I meant with "no side-effect", which wasn't super clear), I figured there are very few chances multiple extensions would actually share references to such newly created lists/sets. Anyway, I can definitely change that to return a `_BlockLevelElements` instance, to be more robust :slightly_smiling_face:

Reply:
Sorry, I missed that this comment was in a specific method which returns a modified object. Yeah, in this case, returning the same object type makes sense.

Reply:
No worries, I wasn't very clear about it. @facelessuser [expressed](https://github.com/Python-Markdown/markdown/pull/1510#discussion_r1973856848) the same concern, so we might want to return a new instance of `_BlockLevelElements`, just to be safe. Let me know what you prefer.

Reply:
I think this is fine as-is. The warning will alert the user that they need to change their code, but in the meantime, we are returning an object of the type that they expect.

Reply:
Alright! Thanks.